src/main/anovos/data_ingest/data_ingest.py

```diff
@@ -41,13 +55,129 @@ def read_dataset(spark, file_path, file_type, file_configs={}):
+        if id_cols:
+            for col in id_cols:
+                if col not in odf.columns:
+                    raise TypeError("Invalid input for id_cols: " + col)
+        if type(threshold_num) != int:
+            raise TypeError("Invalid input for threshold_num")
+        if type(threshold_ratio) != float or threshold_ratio < 0 or threshold_ratio > 1:
+            raise TypeError("Invalid input for threshold_ratio")
+        if (
+            type(threshold_string) != float
+            or threshold_string < 0
+            or threshold_string > 1
+        ):
+            raise TypeError("Invalid input for threshold_string")
+        list_of_cols = list(c for c in odf.columns if c not in id_cols)
+        odf = odf.persist(pyspark.StorageLevel.MEMORY_AND_DISK)
+        odf_recast = odf.select(
+            [F.col(c).cast(T.DoubleType()).alias(c) for c in list_of_cols]
+        )
+        funs_count = [F.count]
+        exprs_count = [f(F.col(c)) for f in funs_count for c in list_of_cols]
+        list_count_recast = (
+            odf_recast.groupby().agg(*exprs_count).rdd.flatMap(lambda x: x).collect()
+        )
+
+        list_count = odf.groupby().agg(*exprs_count).rdd.flatMap(lambda x: x).collect()
+
+        funs_distinct = [F.countDistinct]
+        exprs_distinct = [f(F.col(c)) for f in funs_distinct for c in list_of_cols]
+        list_distinct_count = (
+            odf.select(*exprs_distinct).rdd.flatMap(lambda x: x).collect()
+        )
+        list_ratio = [i / j for i, j in zip(list_distinct_count, list_count)]
+        list_cat_ratio = [i / j for i, j in zip(list_count_recast, list_count)]
+
+        list_schema = []
+        if id_cols:
+            for col in id_cols:
+                list_schema.append(T.StructField(str(col), T.StringType(), True))
+        for k in range(0, len(list_count)):
+            if list_cat_ratio[k] < threshold_string:
+                list_schema.append(
+                    T.StructField(str(list_of_cols[k]), T.StringType(), True)
+                )
+            elif list_count[k] <= threshold_num:
+                if list_ratio[k] > threshold_ratio:
+                    list_schema.append(
+                        T.StructField(str(list_of_cols[k]), T.DoubleType(), True)
+                    )
+                else:
+                    list_schema.append(
+                        T.StructField(str(list_of_cols[k]), T.StringType(), True)
+                    )
+            elif (
+                list_ratio[k] < threshold_ratio
+                or list_distinct_count[k] < threshold_num
+            ):
+                list_schema.append(
+                    T.StructField(str(list_of_cols[k]), T.StringType(), True)
+                )
+
+            else:
+                list_schema.append(
+                    T.StructField(str(list_of_cols[k]), T.DoubleType(), True)
+                )
+        odf.unpersist()
+        full_schema = T.StructType(list_schema)
+        odf = (
```

Comment:
Why are you reading the dataframe again?

Reply:
Can't you just update the dataframe in-memory? Writing to a file and then reading it again seems to be a performance bottleneck.

Reply:
Do you have any data/indication that lazy evaluation of `cast` causes a performance hit? I would be surprised to learn that re-casting in-memory is generally slower than writing and reading from disk.

Reply:
`cast` transformation will definitely run in an instant after you do that operation. However, the actual runtime of this transformation will happen outside the scope of this function, aka whenever an action operation is called. This make the downstream application that use this output much more unpredictable, since in addition to the action operation runtime, some intermediate memory will be created on the spark session, and it will add additional stress on it. Instead, if rereading from the source, the runtime in general will be the same with the total runtime of `cast` and action operation, and there will be no additional stress on spark session, as well as downstream application, which is the purpose of `read_dataset`.